code/modules/integrated_electronics/subtypes/reagents.dm

```diff
@@ -318,6 +323,82 @@
 	complexity = 16
 	spawn_flags = IC_SPAWN_RESEARCH
 
+/obj/item/integrated_circuit/reagent/storage/grinder
+	name = "reagent grinder"
+	desc = "This is reagent grinder.It accepts ref to something and refines it into reagents. Can store up to 100u."
+	icon_state = "blender"
+	extended_desc = ""
+	inputs = list(
+		"target" = IC_PINTYPE_REF,
+		)
+	outputs = list(
+		"volume used" = IC_PINTYPE_NUMBER,
+		"self reference" = IC_PINTYPE_REF
+		)
+	activators = list(
+		"grind" = IC_PINTYPE_PULSE_IN,
+		"on grind" = IC_PINTYPE_PULSE_OUT,
+		"on fail" = IC_PINTYPE_PULSE_OUT
+		)
+	volume = 100
+	power_draw_per_use = 150
+	complexity = 16
+	spawn_flags = IC_SPAWN_RESEARCH
+
+/obj/item/integrated_circuit/reagent/storage/grinder/do_work()
+	if(reagents.total_volume >= reagents.maximum_volume)
+		activate_pin(3)
+		return FALSE
+	var/obj/item/I = get_pin_data_as_type(IC_INPUT, 1, /obj/item)
+	if(istype(I)&&(I.grind_results)&&check_target(I))
+		if(I.on_grind(src) != -1)
+			reagents.add_reagent_list(I.grind_results)
+			if(I.reagents)
+				I.reagents.trans_to(src, I.reagents.total_volume)
+			qdel(I)
+			activate_pin(2)
+			return TRUE
+	activate_pin(3)
+	return FALSE
+
+obj/item/integrated_circuit/reagent/storage/juicer
+	name = "reagent juicer"
+	desc = "This is reagent juicer.It accepts ref to something and refines it into reagents. Can store up to 100u."
+	icon_state = "blender"
+	extended_desc = ""
+	inputs = list(
+		"target" = IC_PINTYPE_REF,
+		)
+	outputs = list(
+		"volume used" = IC_PINTYPE_NUMBER,
+		"self reference" = IC_PINTYPE_REF
+		)
+	activators = list(
+		"juice" = IC_PINTYPE_PULSE_IN,
+		"on juice" = IC_PINTYPE_PULSE_OUT,
+		"on fail" = IC_PINTYPE_PULSE_OUT
+		)
+	volume = 100
+	power_draw_per_use = 150
+	complexity = 16
+	spawn_flags = IC_SPAWN_RESEARCH
+
+/obj/item/integrated_circuit/reagent/storage/juicer/do_work()
+	if(reagents.total_volume >= reagents.maximum_volume)
+		activate_pin(3)
+		return FALSE
+	var/obj/item/I = get_pin_data_as_type(IC_INPUT, 1, /obj/item)
+	if(istype(I)&&check_target(I)&&(I.juice_results))
+		if(I.on_juice() != -1)
```

Comment:
As I said before:
`if(istype(I) && check_target(I) && I.juice_results && I.on_juice() != -1)`
These kinds of conditionals short circuit if it knows earlier that the check as a whole is going to fail.
`check_target()` wont be run if the `istype()` fails. `I.on_juice()` wont be run if there is no `I.juice_results`

Reply:
needed knowlege